Devtrain: Ruby Hi-Lo

donlinOW opened the pull request:
implemented basic features

donlinOW committed:
add project name

donlinOW committed:
update tab space from 4 to 2 spaces

lucy-dement commented on hilo.rb:
```diff
@@ -1,26 +1,28 @@
 # predefined number of guesses allowed
 guesses = 6
 ## computer secretly picks number between 1 and 100 inclusive
 number = rand(1...101)
```
So weird rails syntax things. `1...101` is 1 - 100 but `1..100` also is. The `...` is exclusive and `..` is inclusive, and I need to google it everytime because I forget which is which.

donlinOW replied:
I see, is there a preference for which is used? Perhaps `..` for readability?

lucy-dement commented on hilo.rb:
```diff
@@ -1,26 +1,28 @@
+# Devtrain: Ruby Hi-Lo
+
 # predefined number of guesses allowed
 guesses = 6
```
So it'd be better if this was in a class as it encapsulates all this logic so if you needed to add this to a larger thing(like a games platform) it would be much easier to use elsewhere.

like this
```
class HiLoGame
  def initialize
    @guesses = 6
    @number = ...
  end

  def play
    ..your code here
  end
end

HiLoGame.new.play
```

lucy-dement commented on hilo.rb:
```diff
@@ -1,26 +1,28 @@
+    puts "You win! #{number} was the number!"
+    break
+  # if guess too high, decrement number of guesses left and tell them their guess was too high
+  elsif guess > number
```
You've got some really lovely and clear comments here, but part of the ruby philosophy is writing code that doesn't need comments because it's so obvious what it does. We manage this by naming variables and methods well. For example in this case i'd suggest making a method like this
```
def too_high?(guess)
  ..I'll leave you to figure it out
end
```
so then you can just check
```
elseif too_high?(guess)
```
which is very self explanatory.

Note I've named this method with a `?` on the end, this is just a helpful pattern that shows that this method returns a boolean.

donlinOW committed:
update spacing

lucy-dement commented on hilo.rb:
```diff
@@ -1,26 +1,28 @@
+    puts "Your guess is too high!"
+  # if guess too low, decrement number of guesses left and tell them their guess was too low
+  elsif guess < number
+    guesses -=1
```
Now I'm just being picky 😂 but you have `guesses -= 1` when you could have it outside the if statement. It's also a bit more future proof because if you refactor these if statements you won't have to worry about the incrementing.

donlinOW replied:
done!

lucy-dement commented on hilo.rb:
```diff
@@ -1,30 +1,32 @@
+  elsif guess < number
+    guesses -=1
+    puts "Your guess is too Low!"
+  end
```
Lol, I saw that 😂

donlinOW replied:
:p

donlinOW committed:
refactor random number generation

use ".." for 1-100 inclusive

donlinOW committed:
refactor decrementing of guesses left

moved to outside of if statement, as executed regardless of specific conditon (too high/too low), if user guess is correct this part is skipped due to break statement

donlinOW committed:
add linter (rubocop gem)

donlinOW committed:
make methods for different possible conditions + remove commnets

more easy to read even wo use of additional comments

donlinOW committed:
refactor code to encapsulate game as a class

+ rewrite methods for different conditions (e.g. too low, too high, correct) using instance variables (remove additional number arg)

donlinOW committed:
use single quotes (') instead of double (") for strings

- on recommendation of rubocop linter, more efficient

donlinOW committed:
add top-level documentation comment to describe functionality of class

donlinOW committed:
use double quotation (") as interpolation is needed

donlinOW committed:
use built in ruby methods (.positive?, .zero?)

donlinOW committed:
use guard clause instead of using if statement

donlinOW committed:
remove redundant return statement

lucy-dement commented on hilo.rb:
```diff
@@ -1,30 +1,42 @@
+    puts "You ran out of guesses! The number was #{@number}" if @guesses.zero?
+  end
+
+  def too_high?(guess)
```
These methods should be private so that they can't be accessed outside this class. It helps when working with larger codebases to know which methods are designed to be used by other things and which are strictly internal.

lucy-dement commented on Gemfile:
```diff
@@ -0,0 +1,7 @@
+
+# gem "rails"
+
+gem 'rubocop', require: false
```
This is the really good idea. It means you don't need to think about syntax things too much.

lucy-dement reviewed (changes requested):
This looks great. Just one more small change.

donlinOW committed:
make methods private

private keyword makes all methods defined below it private until explicitly switched to public (or other visibility modifiers)
